### PR TITLE
Major LLM services and UI improvement

### PR DESCRIPTION
**Pull request by GlebSolovev**
### ⏳ Estimate time for LLM services to become available via logging proof-generation requests 

* Implement `GenerationsLogger`
  * capable of thoroughly logging all LLM service generation requests in a human-readable way, which can be useful for debugging too;
  * with support of censorship for sensitive data (for example, API keys).
* Implement a default algorithm to estimate the time for LLM service to become available.

### 🕵️ Support proper logging and error handling inside `LLMService`-s

* Introduce hierarchy for `LLMService` errors: `LLMServiceError`, `ConfigurationError`, `RemoteConnectionError`, `GenerationFailedError`.
* Handle these errors properly inside `LLMService` implementation: 
  * throw different errors for different causes;
  * repack errors where needed.
* Implemented proper logging based on the errors thrown: with both `EventsLogger` and `GenerationsLogger`.
* Introduce `LLMServiceRequest` to make data transfer coherent between different logic modules.
* Design and implement interaction between `LLMService`-s and UI to report errors.

### 🏯 Rework `LLMService` arhcitecture

*  Introduce and implement `LLMServiceInternal` that:
    * separates and hides the actual implementation of the interaction with a model;
    * restricts the visibility of some methods and properties;
    * provides a convenient way to support new services easily via wrapper methods.
* Rework facade classes and methods of `LLMService` to make them both concise and powerful.
* Design and support complete and powerful typing for `LLMService` classes based on recursive generics.
  * That makes any further development involving `LLMService`-s safer and easier. 
* Document core `LLMService` architecture.

### ✅ Test everything

* Test the LLM services module completely:
    * including facade, core and utility functionality;
    * via both unit and integration tests, with the use of stress testing and asynchronous behaviour;
    * in total, about 150 new tests.
* Improve test infrastructure in general: 
    * introduce a lot of utility functions and mocks;
    * structure them properly. 
* Fix minor bugs and make minor improvements detected by tests.

### 🚀 Fix & improve CI

* Fix the issue with building a test Coq project resource.
* Thereby, make previously "non-CI" tests successfully pass on CI.
* Set up CI debugging for future needs.
  * Now it is possible to connect to the running GitHub-Actions worker by SSH and debug the environment. This mode activates only in case any task fails. 
* Enable launching CI manually.
* Double the speed of CI by setting up caches for `ocaml` and `opam` dependencies.

### 🤝 Improve LLM services' parameters: their naming, transparency, description

* Introduce `modelId` to distinguish a model identifier from the name of an OpenAI / Grazie model.
* Rename `newMessageMaxTokens` to `maxTokensToGenerate` for greater clarity.
* Add `defaultChoices` to `ModelParams` to make its resolution more transparent in the code.
* Update the settings description, and make it more user-friendly.

### 🕊 Rework and significantly improve settings validation

* Improve Ajv schemas for the parameters.
* Refactor validation errors to one `SettingsValidationError`,
    * which supports hints for the settings to open.
* Move all input parameters validation to one place and make it coherent.
* Add some general checks:
    *  input models have unique `modelId`-s;
    * there are models to generate proofs with.
* Design and implement a powerful parameters resolver framework that:
    * makes it possible for LLM services to resolve input parameters (with overrides and defaults) and validate them;
    * makes it possible for developers to specify parameters resolution in a declarative way;
    * provides great compile-time type checks for developers;
    * is well-documented.
* Significantly improve messages shown to users:
    * improve existing ones;
    * support messages about parameter resolution: both failures and unexpected overrides.
* Decipher Ajv errors and make them clear both in UI and for debugging.

### 💚🤖 Improve interaction with OpenAI

* Handle OpenAI runtime errors and repacked them into clear configuration errors, shown to users.
* Specify default token parameters for `OpenAIService`:
    * gather values for various models from different sources;
    * implement an algorithm to resolve `maxTokensToGenerate` appropriately.

### 🌿 Improve code quality

* Split `CoqPilot` into smaller pieces.
* Make throwing `Error` coherent.
* Get rid of unnecessary `any`-s.
* Output values coherently.
* Make an order of LLM services coherent.
* Improve linters' configuration.

### CODE REVIEW

File: src/extension/configReaders.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+    if (openAiUserParams.some((params) => params.apiKey === "None")) {
+        throw buildApiKeyError("Open Ai", "openAi");
+    }
+    if (grazieUserParams.some((params) => params.apiKey === "None")) {
+        throw buildApiKeyError("Grazie", "grazie");
+    }
+}
+
+function validateModelsArePresent<T>(allModels: T[]) {
```
Comment (K-dizzled) [inline src/extension/configReaders.ts:209]: I see you bravely covered a trillion cases of settings validation.

File: src/extension/configReaders.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        }
+
+        if (resolutionResult.resolved !== undefined) {
+            resolvedParamsList.push(resolutionResult.resolved);
+        }
+    }
+    return resolvedParamsList;
+}
+
+function buildResolutionHistory(
```
Comment (K-dizzled) [inline src/extension/configReaders.ts:271]: Elaborate on the task of this function please.

File: src/llm/llmServiceErrors.ts
Comment (K-dizzled) [inline src/llm/llmServiceErrors.ts:1]: Neat

File: src/llm/llmServices.ts [hunk 3]
```diff
@@ -1,11 +1,51 @@
...
+): LLMService<UserModelParams, ModelParams>[] {
+    return [
+        llmServices.predefinedProofsService,
+        llmServices.openAiService,
+        llmServices.grazieService,
+        llmServices.lmStudioService,
+    ];
+}
+
+export function switchByLLMServiceType<T>(
```
Comment (K-dizzled) [inline src/llm/llmServices.ts:31]: Elaborate

File: src/llm/llmServices/utils/generationsLogger/loggerRecord.ts [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+
+    private estimatedTokensToString(): string {
+        return `${this.estimatedTokens!.maxTokensInTotal} = ${this.estimatedTokens!.messagesTokens} (chat messages) + ${this.estimatedTokens!.maxTokensToGenerate} (max to generate)`;
+    }
+
+    protected static escapeNewlines(text: string): string {
+        return text.replace("\n", "\\n");
+    }
+
+    static deserealizeFromString(rawRecord: string): [LoggerRecord, string] {
```
Comment (K-dizzled) [inline src/llm/llmServices/utils/generationsLogger/loggerRecord.ts:97]: How often is this being called? Doesn't it bring to much overhead in computational costs?

File: src/test/commonTestFunctions/coqFileParser.ts
Comment (K-dizzled) [inline src/test/commonTestFunctions/coqFileParser.ts:1]: You have moved functions that work as `solve` functions from the particular tests to the special folder. I am not sure that's a great idea. Their semantic is not that clear and, I believe, they are used pretty much once. Or maybe I've not seen the new tests enough.

File: src/test/commonTestFunctions/delay.ts
Comment (K-dizzled) [inline src/test/commonTestFunctions/delay.ts:1]: Fanstastic

File: .github/workflows/coqpilot.yml
Comment (K-dizzled) [inline .github/workflows/coqpilot.yml:1]: Where do you build the coq project in this workflow?

File: src/test/index.ts
Comment (K-dizzled) [inline src/test/index.ts:1]: Why now Error is created without `new` keyword?

File: src/test/llm/llmServices/llmService/availabilityEstimator.test.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import {
+    Time,
+    nowTimestampMillis,
+    time,
+    timeToMillis,
+    timeToString,
+    timeZero,
+} from "../../../../llm/llmServices/utils/time";
+
+suite("[LLMService] Test default availability estimator", () => {
```
Comment (K-dizzled) [inline src/test/llm/llmServices/llmService/availabilityEstimator.test.ts:17]: What is the semantics of the [...] in tests?

File: src/test/llm/llmSpecificTestUtils/withMockLLMService.ts
Comment (K-dizzled) [inline src/test/llm/llmSpecificTestUtils/withMockLLMService.ts:1]: Neat

**Review by K-dizzled — CHANGES_REQUESTED** [review]

File: .github/workflows/coqpilot.yml
[the remark this answers is not in the document]
Comment (GlebSolovev) [reply .github/workflows/coqpilot.yml:1]: Yup, made it pretty much the same as you did in `ai_agents_server` and also used it on CI

File: src/llm/llmServices/utils/generationsLogger/loggerRecord.ts (hunk 4 above)
Comment (GlebSolovev) [reply src/llm/llmServices/utils/generationsLogger/loggerRecord.ts:97]: It is a good question to be answered indeed, including the docs. I left a note at `GenerationsLogger`'s docs about the possible performance overhead and how to resolve the issue if it arises in practice one day.

No worries for now though, neither tests nor I didn't experience any problems, on the opposite, only +vibes from working with human-readable logs during debugging 🕊

**Pull request merged**